File: server/src/hephaestus/agent_bridge/framing.py
"""LF-delimited JSON-RPC 2.0 framing — Python mirror of ``agent/src/framing.ts``.

The two implementations share one wire contract so that a frame emitted by
either side is parsed identically by the other, and :func:`encode_frame` produces
**byte-for-byte identical output** to the TypeScript ``encodeFrame`` for the same
value (cross-language golden fixture). Byte parity relies on a canonical JSON
serialization: recursively sorted object keys, compact separators, and raw
(non-ASCII-escaped) UTF-8.

Framing rules (architecture §5):

* Frames are UTF-8 JSON objects terminated by a single ``\\n``; protocol stdout
  never carries anything else (logs go to stderr).
* The decoder is **incremental**: it aborts as soon as an in-progress frame
  exceeds :data:`MAX_FRAME_BYTES`, without buffering the whole oversized frame.
* :func:`encode_frame` refuses to emit an oversized frame (outbound guard).
"""

from __future__ import annotations

import json
from typing import Any

from .limits import MAX_FRAME_BYTES

__all__ = [
    "FrameDecoder",
    "FrameError",
    "FrameTooLargeError",
    "canonical_json",
    "encode_frame",
]

_NL = 0x0A


class FrameError(Exception):
    """A framing-layer violation. ``code`` is a stable machine token."""

    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code
        self.message = message


class FrameTooLargeError(FrameError):
    """A single frame exceeded :data:`MAX_FRAME_BYTES` in either direction."""

    def __init__(self, observed_bytes: int, max_bytes: int) -> None:
        super().__init__(
            "frame_too_large",
            f"frame exceeds {max_bytes} bytes (observed at least {observed_bytes})",
        )
        self.observed_bytes = observed_bytes
        self.max_bytes = max_bytes


def canonical_json(value: Any) -> str:
    """Canonical, cross-language-stable JSON text.

    Sorted keys + compact separators + ``ensure_ascii=False`` match the
    TypeScript ``canonicalJson`` byte-for-byte for JSON values that avoid
    floating-point formatting differences (frames use only strings, integers,
    booleans, null, arrays, and objects).
    """
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def encode_frame(frame: dict[str, Any]) -> bytes:
    """Serialize one frame to canonical JSON + ``\\n``; enforce the outbound cap."""
    payload = (canonical_json(frame) + "\n").encode("utf-8")
    if len(payload) > MAX_FRAME_BYTES:
        raise FrameTooLargeError(len(payload), MAX_FRAME_BYTES)
    return payload


class FrameDecoder:
    """Incremental LF framer with a hard per-frame byte cap.

    Feed raw ``bytes`` via :meth:`push`; it returns the list of complete frame
    payloads (without the trailing newline) discovered so far, retaining any
    partial trailing line for the next call. Blank lines are skipped. As soon as
    an in-progress line exceeds the cap — even before its newline arrives — a
    :class:`FrameTooLargeError` is raised (fail closed); the caller tears the
    connection down rather than attempting to resynchronize.
    """

    def __init__(self) -> None:
        self._buf = bytearray()

    def push(self, chunk: bytes) -> list[bytes]:
        frames: list[bytes] = []
        start = 0
        n = len(chunk)
        while start < n:
            nl = chunk.find(_NL, start)
            if nl == -1:
                self._buf += chunk[start:]
                if len(self._buf) > MAX_FRAME_BYTES:
                    raise FrameTooLargeError(len(self._buf), MAX_FRAME_BYTES)
                break
            self._buf += chunk[start:nl]
            if len(self._buf) > MAX_FRAME_BYTES:
                raise FrameTooLargeError(len(self._buf), MAX_FRAME_BYTES)
            line = bytes(self._buf)
            self._buf.clear()
            start = nl + 1
            if line.strip():
                frames.append(line)
        return frames

    @property
    def buffered(self) -> int:
        """Bytes of a partial (unterminated) frame currently held."""
        return len(self._buf)

File: agent/src/framing.ts
// LF-delimited JSON-RPC 2.0 framing — TypeScript mirror of framing.py.
//
// canonicalJson() + encodeFrame() produce BYTE-FOR-BYTE identical output to the
// Python side for the shared golden fixture (ASCII keys; string/int/bool/null/
// array/object values), because both use recursively sorted object keys, compact
// separators, and raw (non-escaped) UTF-8. The decoder is incremental: it aborts
// as soon as an in-progress frame exceeds the 64 MiB cap, without buffering the
// whole oversized frame.

import { WIRE_MAX_FRAME_BYTES } from "./limits.js";

export type JsonValue = null | boolean | number | string | JsonValue[] | { [k: string]: JsonValue };

const NL = 0x0a;

export class FrameError extends Error {
  constructor(
    readonly code: string,
    message: string,
  ) {
    super(message);
    this.name = "FrameError";
  }
}

export class FrameTooLargeError extends FrameError {
  constructor(
    readonly observedBytes: number,
    readonly maxBytes: number,
  ) {
    super(
      "frame_too_large",
      `frame exceeds ${maxBytes} bytes (observed at least ${observedBytes})`,
    );
    this.name = "FrameTooLargeError";
  }
}

/**
 * Canonical, cross-language-stable JSON text: object keys sorted, compact
 * separators, raw UTF-8. Matches Python `json.dumps(sort_keys=True,
 * separators=(",",":"), ensure_ascii=False)` for the fixture value domain.
 */
export function canonicalJson(value: JsonValue): string {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return "[" + value.map(canonicalJson).join(",") + "]";
  }
  const obj = value as { [k: string]: JsonValue };
  const keys = Object.keys(obj).sort();
  const members = keys.map((k) => JSON.stringify(k) + ":" + canonicalJson(obj[k] as JsonValue));
  return "{" + members.join(",") + "}";
}

/** Serialize one frame to canonical JSON + "\n"; enforce the outbound cap. */
export function encodeFrame(frame: { [k: string]: JsonValue }): Buffer {
  const buf = Buffer.from(canonicalJson(frame) + "\n", "utf8");
  if (buf.length > WIRE_MAX_FRAME_BYTES) {
    throw new FrameTooLargeError(buf.length, WIRE_MAX_FRAME_BYTES);
  }
  return buf;
}

/**
 * Incremental LF framer with a hard per-frame byte cap. Feed raw chunks via
 * push(); it returns the complete frame payloads (without the trailing newline)
 * found so far, retaining any partial trailing line for the next call. Blank
 * lines are skipped. Exceeding the cap — even before a newline arrives — throws
 * FrameTooLargeError (fail closed): the caller tears the connection down.
 */
export class FrameDecoder {
  private chunks: Buffer[] = [];
  private len = 0;

  push(chunk: Buffer): Buffer[] {
    const frames: Buffer[] = [];
    let start = 0;
    const n = chunk.length;
    while (start < n) {
      const nl = chunk.indexOf(NL, start);
      if (nl === -1) {
        const rest = chunk.subarray(start);
        this.chunks.push(rest);
        this.len += rest.length;
        if (this.len > WIRE_MAX_FRAME_BYTES) {
          throw new FrameTooLargeError(this.len, WIRE_MAX_FRAME_BYTES);
        }
        break;
      }
      const part = chunk.subarray(start, nl);
      this.chunks.push(part);
      this.len += part.length;
      if (this.len > WIRE_MAX_FRAME_BYTES) {
        throw new FrameTooLargeError(this.len, WIRE_MAX_FRAME_BYTES);
      }
      const line = Buffer.concat(this.chunks);
      this.chunks = [];
      this.len = 0;
      start = nl + 1;
      if (line.toString("utf8").trim().length > 0) {
        frames.push(line);
      }
    }
    return frames;
  }

  get buffered(): number {
    return this.len;
  }
}
